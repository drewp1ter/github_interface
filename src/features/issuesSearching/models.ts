export interface IIssuesRequest {
  userName: string
  repoName: string
  issuesState: 'all' | 'open' | 'closed'
}

export interface IIssue {
  id: number
  created_at: string
  title: string
  body: string
  number: number
  user: {
    login: string
    avatar_url: string
    html_url: string
  }
}

export interface IIssues {
  userName: string,
  repoName: string,
  payload: IIssue[]
}

export const createIssue = (props?: IIssue): IIssue => ({
  id: -1,
  created_at: '',
  title: '',
  body: '',
  number: -1,
  user: {
    login: '',
    avatar_url: '',
    html_url: ''
  },
  ...props
})