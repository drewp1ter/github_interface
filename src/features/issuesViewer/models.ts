export interface IIssuesRequest {
  userName: string
  repoName: string
}

export interface IIssue {
  id: number
  created_at: string
  title: string
  body: string
  number: number
  user: {
    login: string
    avatar_url: string
    html_url: string
  }
}