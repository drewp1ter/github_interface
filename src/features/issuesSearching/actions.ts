import { createAsyncAction } from 'typesafe-actions'
import { IIssuesRequest, IIssues } from './models'

import * as types from './actionTypes'

export const fetchIssues = createAsyncAction(types.FETCH_ISSUES_REQUEST, types.FETCH_ISSUES_SUCCESS, types.FETCH_ISSUES_FAILURE)<
  IIssuesRequest,
  IIssues,
  IRequestError
>()
