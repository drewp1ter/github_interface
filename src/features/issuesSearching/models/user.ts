export interface IUserDTO {
  id: number
  login: string
  avatar_url: string
  html_url: string
}

export interface IUser {
  id: number
  login: string
  avatarUrl: string
  htmlUrl: string
}

export class User implements IUser {

  public static create(dto: IUserDTO): IUser {
    return new User(dto.id, dto.login, dto.avatar_url, dto.html_url)
  }
  constructor(public id: number = -1, public login: string = '', public avatarUrl: string = '', public htmlUrl: string = '') {}
}
