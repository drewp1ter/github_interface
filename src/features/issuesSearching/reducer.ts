import { combineReducers } from 'redux'
import { ActionType } from 'typesafe-actions'

import { IIssues } from './models'
import * as actions from './actions'
import * as types from './actionTypes'

export interface IssuesSearchingState {
  readonly issues: IIssues
  readonly fetching: boolean
  readonly error: IRequestError
}

export type IssuesSearchingAction = ActionType<typeof actions>

export default combineReducers<IssuesSearchingState, IssuesSearchingAction>({
  issues: (state = { userName: '', repoName: '', payload: [] }, action) =>
    action.type === types.FETCH_ISSUES_SUCCESS ? action.payload : state,
  fetching: (state = false, action) =>
    action.type === types.FETCH_ISSUES_REQUEST || (state && !(types.FETCH_ISSUES_SUCCESS || types.FETCH_ISSUES_FAILURE)),
  // tslint:disable-next-line: variable-name
  error: (_state, action) => (action.type === types.FETCH_ISSUES_FAILURE ? action.payload : { message: '', status: 0 }),
})
