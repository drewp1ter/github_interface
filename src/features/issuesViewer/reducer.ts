import { combineReducers } from 'redux'
import { ActionType } from 'typesafe-actions'

import { IIssue } from './models'
import * as actions from './actions'
import * as types from './actionTypes'

export interface IIssuesViewerState {
  readonly issues: IIssue[]
  readonly fetching: boolean
  readonly error: RequestError
}

export type IssuesViewerAction = ActionType<typeof actions>

export default combineReducers<IIssuesViewerState, IssuesViewerAction>({
  issues: (state = [], action) => action.type === types.FETCH_ISSUES_SUCCESS ? action.payload : state,
  fetching: (state = false, action) => action.type === types.FETCH_ISSUES_REQUEST || state && !(types.FETCH_ISSUES_SUCCESS || types.FETCH_ISSUES_FAILURE),
  error: (state, action) => action.type === types.FETCH_ISSUES_FAILURE ? action.payload : { message: '' }
})