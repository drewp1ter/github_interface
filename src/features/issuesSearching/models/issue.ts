import { User, IUser, IUserDTO } from './user'

export interface IIssuesRequest {
  userName: string
  repoName: string
  issuesState: 'all' | 'open' | 'closed'
}

export interface IIssueDTO {
  id: number
  created_at: string
  title: string
  body: string
  number: number
  state: 'closed' | 'open',
  user: IUserDTO
}

export interface IIssue {
  id: number
  createdAt: string
  createdAtFormated: string
  title: string
  body: string
  number: number
  state: 'open' | 'closed'
  user: IUser
}

export class Issue implements IIssue {

  get createdAtFormated(): string {
    const date = new Date(this.createdAt)
    return date.toDateString()
  }

  public static create(dto: IIssueDTO): IIssue {
    return new Issue(dto.id, dto.created_at, dto.title, dto.body, dto.number, dto.state, User.create(dto.user))
  }
  constructor(
    public id: number = -1,
    public createdAt: string = '',
    public title: string = '',
    public body: string = '',
    public number: number = -1,
    public state: 'open' | 'closed' = 'open',
    public user = new User()
  ) {}
}

export interface IIssues {
  userName: string
  repoName: string
  payload: IIssue[]
}
