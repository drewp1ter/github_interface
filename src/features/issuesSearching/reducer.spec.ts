import reducer, { IssuesSearchingState } from './reducer'
import * as actions from './actions'
import { Issue } from './models'

const getInitialState = (initial?: Partial<IssuesSearchingState>) => reducer(initial as IssuesSearchingState, {} as any)

describe('IssuesSearching Stories', () => {
  const initialState = getInitialState()

  it('should match a snapshot', () => {
    expect(initialState).toMatchSnapshot()
  })

  it('issues request fetching true', done => {
    const state = reducer(initialState, actions.fetchIssues.request({ userName: '', repoName: '', issuesState: 'all' }))
    expect(state).toEqual({
      ...initialState,
      fetching: true,
    })
    done()
  })

  it('issues fetching success', done => {
    const issues = {
      userName: 'mufdvr',
      repoName: 'test',
      payload: [new Issue()],
    }
    expect(initialState.issues.payload).toHaveLength(0)
    const state = reducer(initialState, actions.fetchIssues.success(issues))
    expect(state).toEqual({
      ...initialState,
      issues,
    })
    done()
  })

  it('issue fetching error', done => {
    const error = {
      message: 'Not found',
      status: 404,
    }
    const state = reducer(initialState, actions.fetchIssues.failure(error))
    expect(state.issues.payload).toHaveLength(0)
    expect(state).toEqual({
      ...initialState,
      error,
    })
    done()
  })
})
