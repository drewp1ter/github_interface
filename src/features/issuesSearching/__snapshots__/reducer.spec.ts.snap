// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuesSearching Stories should match a snapshot 1`] = `
Object {
  "error": Object {
    "message": "",
    "status": 0,
  },
  "fetching": false,
  "issues": Object {
    "payload": Array [],
    "repoName": "",
    "userName": "",
  },
}
`;
